src/openms/source/METADATA/AbsoluteQuantitationStandards.cpp

```diff
@@ -28,34 +28,80 @@
 // ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 //
 // --------------------------------------------------------------------------
-// $Maintainer: Douglas McCloskey $
-// $Authors: Douglas McCloskey $
+// $Maintainer: Douglas McCloskey, Pasquale Domenico Colaianni $
+// $Authors: Douglas McCloskey, Pasquale Domenico Colaianni $
 // --------------------------------------------------------------------------
 
 #include <OpenMS/METADATA/AbsoluteQuantitationStandards.h>
 
-#include <cstddef> // for size_t & ptrdiff_t
-#include <vector>
-#include <string>
-
 namespace OpenMS
 {
-    
-    AbsoluteQuantitationStandards::AbsoluteQuantitationStandards()
-    {
-    }
-    
-    AbsoluteQuantitationStandards::~AbsoluteQuantitationStandards()
+  void AbsoluteQuantitationStandards::mapComponentsToConcentrations(
+    const std::vector<AbsoluteQuantitationStandards::runConcentration>& run_concentrations,
+    const std::vector<FeatureMap>& feature_maps,
+    std::map<String, AbsoluteQuantitationStandards::featureConcentration>& components_to_concentrations
+  ) const
+  {
+    components_to_concentrations.clear();
+    for (const FeatureMap& fmap : feature_maps)
     {
+      if (!fmap.metaValueExists("sample_name"))
+      {
+        throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "The FeatureMap misses the 'sample_name' MetaValue information.");
+      }
+      const String sample_name = fmap.getMetaValue("sample_name");
+      for (const Feature& feature : fmap)
+      {
+        const std::vector<Feature>& subordinates = feature.getSubordinates();
+        if (subordinates.size() != 2)
+        {
+          continue;
+        }
+        const Feature& f1 = subordinates[0];
+        const Feature& f2 = subordinates[1];
+        std::vector<AbsoluteQuantitationStandards::runConcentration>::const_iterator it;
+        it = std::find_if(
+          run_concentrations.begin(), run_concentrations.end(), [&sample_name, &f1, &f2] (AbsoluteQuantitationStandards::runConcentration run)
+          {
+            if (!(f1.metaValueExists("native_id") && f2.metaValueExists("native_id")))
+            {
+              //throw Exception::MissingInformation(__FILE__, __LINE__, OPENMS_PRETTY_FUNCTION, "The feature misses the 'native_id' MetaValue information.");
+              LOG_DEBUG << "The feature misses the 'native_id' MetaValue information.\n";
+              return false;
+            }
+            return sample_name == run.sample_name &&
```

Comment:
The assumption that both the IS and component will be in the same `feature` will often be violated.  See https://github.com/OpenMS/OpenMS/blob/develop/src/openms/source/ANALYSIS/QUANTITATION/AbsoluteQuantitation.cpp#L324.  Often it will be the case that the IS and component are in the same feature, but this should not be assumed, and the method should fall back to looking through each feature/subfeature for the component and IS similar to what I did in the code above.